### ISSUE TITLE
Avoid lossy buffering in #[serde(flatten)]

### ISSUE THREAD

**Issue by sfackler** [serde-rs/serde#2186]
The `Deserialize` implementation of a type with a field annotated `#[serde(flatten)]` currently buffers up all of the unknown map entries in memory, and then deserializes the flattened value from them at the end. This approach has a couple of downsides:

1. It causes the deserializer to allocate when that would normally not be the case.
2. It confuses "smart" deserializers - things like serde_ignored_value or serde_json's error position tracking don't work properly.
3. It can lose out on special casing in deserializers based on hinting.

Fortunately, there's alternate approach that avoids all of these issues! Instead of buffering, we create a custom `MapAccess` which intercepts fields that we want to handle in the top-level type, and then returning the remainder to the nested deserialize implementation: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=96546d7ba882b70042268a0b48de6286

**Comment by jonasbb** [comment]
This is a very interesting idea. 
I have a couple of remarks / incompatibility observations. Maybe some of them can be resolved.

Currently, the `Foo` type ultimately decides if a field name is known or unknown. This leads to an incompatibility currently, if `deny_unknown_fields` is added to `Foo`. With the normal `flatten` extra fields will be ignored, since `ExtendedFoo` allows them.
https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=4e249bb1961ff44945606ed6bdcdab7d

I think that can be fixed by adding an `Ignore` variant to `ExtendedFooFields`, similar to the current derive. An unknown field error from the seed of `ExtendedFooFieldsVisitor` must be converted into the `Ignore` variant, but all other error likely should remain. Inspecting the errors might be complicated, since they are very opaque.

It is unclear to me how this could be extended to more than one flattened field.
https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=7451b4980894d2ddda33b7ce3fc0aa9c
An additional complication here is the fact, that the same value can be deserialized multiple times. A map visitor (i.e., `Value`) will not consume the fields from the internal buffer, but a struct visitor (i.e., `Sub`) will. This leads the second `Value` to share all fields with the first except those consumed by the `Sub` visitor.
<details>
<summary>Output</summary>

```text
[src/main.rs:27] d = Data {
    f0: Object({
        "x": String(
			"foo",
        ),
        "y": String(
            "bar",
        ),
        "z": String(
            "baz",
        ),
    }),
    a: 123,
    f1: Sub {
        y: "bar",
    },
    b: 456,
    f2: Object({
        "x": String(
            "foo",
        ),
        "z": String(
            "baz",
        ),
    }),
}
```

</details>

**Comment by sfackler** [comment]
> An unknown field error from the seed of `ExtendedFooFieldsVisitor` must be converted into the `Ignore` variant, but all other error likely should remain. Inspecting the errors might be complicated, since they are very opaque.

We should be able to handle this by making a custom error type used by the MapAccess implementation that intercepts unknown field errors and delegates the rest to the inner MapAccess's error type.

> It is unclear to me how this could be extended to more than one flattened field.

Wow, I was not aware that was a thing you could do! I think you are stuck buffering there to allow the key value to be deserialized multiple times unfortunately. We could continue using the old logic in that case.

**Comment by sfackler** [comment]
Hmm - we can't actually filter the errors since the error will terminate the deserialization of the inner value entirely...

**Comment by sfackler** [comment]
On the other hand, the docs on serde.rs do explicitly state that `flatten` isn't compatible with `deny_unknown_fields`: https://serde.rs/field-attrs.html#flatten.

**Comment by sfackler** [comment]
Here's a working implementation of that approach: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=b9b5b874c6f4c93d6d82c1bed4e6451e. If you did want to allow unknown fields in `ExtendedFoo`, we'd codegen the extra `Ignored` variant for `ExtendedFooFields` and return that instead of an error in `ExtendedFooFieldsSeed`.

**Comment by Mingun** [comment]
I think, that using a json is a bad example to demonstrate absence of "lossy deserialization", because JSON contains some information about value type. Try some untyped format, like XML (but there are no good XML serde adapter, so using some toy format would be better)

**Comment by Lucretiel** [comment]
I coincidentally made a fully working implementation of this this evening. You don't need to depend at all on looking for particular deserialize errors / missing_field notifications, because you already know ahead of time what the names of the fields are that you're interested in. You can further minimize codegen by creating a new, private trait (I called mine `KeyCapture`):

```rust
pub enum KeyOutcome<T> {
    Accepted(T),
    Rejected,
}

pub trait KeyCapture<'de> {
    type Token;

    fn expecting(&self, formatter: &mut std::fmt::Formatter) -> std::fmt::Result;

    fn try_send_key<E>(&mut self, key: &[u8]) -> Result<KeyOutcome<Self::Token>, E>
    where
        E: de::Error;

    fn send_value<D>(&mut self, token: Self::Token, value: D) -> Result<(), D::Error>
    where
        D: de::Deserializer<'de>;
}
```

In the struct macro codegen, you implement this trait such that it "rejects" keys that are not part of the struct; these keys can then be transparently forwarded to the type being flattened. Once you have this trait, you can use a series of private adapter types to deserialize the inner flattened field while simultaneously sending data into the `KeyCapture`. Some snippets:

```rust
// From MapAccess:
    fn next_key_seed<K>(&mut self, mut seed: K) -> Result<Option<K::Value>, Self::Error>
    where
        K: de::DeserializeSeed<'de>,
    {
        // We're extracting a key for the type. Try to send keys to the `KeyCapture`,
        // and the first key that is rejected by the `KeyCapture` are is instead to the
        // `seed`
        loop {
            seed = match self.map.next_key_seed(ExtractKeySeed {
                seed,
                capture: &mut self.capture,
            })? {
                None => {
                    self.drained = true;
                    return Ok(None);
                }
                Some(ExtractKeySeedResult::Deserialized(value)) => return Ok(Some(value)),
                Some(ExtractKeySeedResult::Captured { seed, token }) => {
                    self.map.next_value_seed(ExtractValueSeed {
                        token,
                        capture: &mut self.capture,
                    })?;

                    seed
                }
            };
        }
    }

// From Visitor:
    fn visit_str<E>(self, v: &str) -> Result<Self::Value, E>
    where
        E: de::Error,
    {
        match self.capture.try_send_key(v.as_bytes())? {
            KeyOutcome::Accepted(token) => Ok(ExtractKeySeedResult::Captured {
                seed: self.seed,
                token,
            }),
            KeyOutcome::Rejected => self
                .seed
                .deserialize(v.into_deserializer())
                .map(ExtractKeySeedResult::Deserialized),
        }
    }


```

**Comment by Lucretiel** [comment]
Completed implementation for consideration: https://github.com/Lucretiel/serde-bufferless

**Comment by Mingun** [comment]
I think it is better to fix this generically, not just for the case with one flattened field, because in that case it creates a friction when users realising that small change in definitions of their types will broke deserialization.

To achieve that, changes only on deserializer side is not enough. I work on this problem, and you can see my unfinished work at https://github.com/serde-rs/serde/compare/master...Mingun:flatten. The actual changes begins from 1071b39e63077789beb9e55af7384462db72a3bd, the other is a prepare work.

The key idea in that `#[derive(Deserialize)]` implements additional trait `DeserializeEmbed` on types that could be flattened. This approach solves many problems:
- you no longer can add `#[serde(flatten)]` on fields of non-flattanable types, such as `String`, because `DeserializeEmbed` is not implemented for strings (see eac15662a5e045918e93e64b5a7215fe12699307)
- you now deserialize each field only once (not consuming flattened fields from deserializer [is a mistake](https://github.com/serde-rs/serde/issues/1909))
- you can pass flattened field to any level (when you flatten in flatten in flatten...)
- you no longer need a hack for support `Option` in untagged enums: https://github.com/serde-rs/serde/blob/c3ce2c934a9123b5a309256d65cca6b6d19466ec/serde/src/de/mod.rs#L1672-L1679
- I expect that almost all [flatten problems](https://github.com/serde-rs/serde/issues?q=flatten+is%3Aopen+is%3Aissue) will be solved by this

I do not remember the exact status of readiness of this, but I remember that only one of the enum representation options remained to be implemented (seems untagged representation).

Chosen approach has downsides, but their are not so big:
- the main downside is that flattenable/non-flattenable is a trait of _a type_, not _a deserialize operation_, but `DeserializeEmbed` derivered on operation basis. This is the fundamental restriction, because I trying to make this invasive as minimal as I possible and I almost approached that: except two scenarios, described below the end-user don't need to change in their code nothing
- strictly speaking, this is a breaking changes, because your code could stop to compile after this change. But serde anyway [doesn't follow](https://github.com/alexcrichton/cc-rs/issues/601#issuecomment-912492901) semver, so the users should be prepare for that (of course, I supposed to include this only in serde 2 and start to follow semver).

The two scenarios is following:
1. User implement `Deserialize` manually _and_ that type is flattened into another type. I think this is very rare case
2. User try to flatten field of type that cannot be flattened, such of `String` example above. This is in any case an error in the user code that anyway should be fixed, because currently such types will _always_ deserialized with an error which, obviously, not what the user want. Unfortunately, because `DeserializeEmbed` implemented only in `#[derive(Deserialize)]`, having only `#[derive(Serialize)]` on the type still allow you to create an erroneous type. I don't know, is it possible to automatically implement some trait twice from different derives without ambiguity, which is by good required here

**Comment by RReverser** [comment]
Ha, reinvented the same wheel (deserializing with just one flat field without buffers) recently and only now discovered this thread. I suppose I should try to switch to serde-bufferless linked above instead.

**Comment by RReverser** [comment]
@Lucretiel Any chance you'd want to add a proc-macro to your repo and publish it to crates.io?

**Comment by palant** [comment]
I’ve taken a similar approach to @Mingun, this issue cannot really be solved without API changes. The deserializers need to expose what fields they can take, and they need to accept fields individually rather than an entire map. I’ve added a `DeserializeMap` trait extending `Deserialize` for map-based types as well as the corresponding `MapVisitor` trait which actually collects the values:

```rust
pub trait DeserializeMap<'de>: Deserialize<'de> {
    type Visitor: MapVisitor<'de, Value = Self> + Sized;
    fn visitor() -> Self::Visitor;
}

pub trait MapVisitor<'de> {
    type Value;
    fn accepts_field(field: &str) -> bool;
    fn list_fields(list: &mut Vec<&'static str>);
    fn visit_field<D>(&mut self, field: &str, deserializer: D) -> Result<(), D::Error>
    where
        D: serde::de::Deserializer<'de>;
    fn finalize<E>(self) -> Result<Self::Value, E>
    where
        E: serde::de::Error;
}
```

The `Deserialize` trait can be implemented generically (as in: blanket implementation would have been possible if not for foreign types) for any data structure implementing `DeserializeMap`. And a `DeserializeMap` implementation can delegate processing of individual keys to other `DeserializeMap` implementations as needed.

Implementation here: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=3e86581a385b2922044efb5156c30745

I’ve got derive macros implemented as well but these are specialized for my use case.

**Comment by cbeck88** [comment]
In another project, I also needed to find a way to make `serde(flatten)` work without internal buffering.

The solution that I settled on is, where `serde_derive` code-gens for each struct a loop which looks roughly like:

```rust
while let Some(key) = map_access.next_key()? {
   match key => {
       "field_name" => field_name = Some(map_access.next_value());
        ...
        _unknown_ => { ... handle unknown key ... }
   }
}
```

my code instead organizes this as a state machine, so the local variables are members of the state machine, and the state machine implements this trait:

https://github.com/cbeck88/conf-rs/blob/fa74ae001525f9293fe803020289e34676a85cbd/src/conf_serde/state_machine.rs#L44

```rust
/// In order to implement serde(flatten) properly, we need to allow for the idea that:
///
/// * struct A contains two flattened children struct B and struct C
/// * some fields are produced by serde::de::MapAccess<'de> which go to A, then which go to B,
///   then which go to C, then which go back to B again, then which go back to C, etc.
///
/// This means that the code we generate for B cannot take a deserializer or similar and generate
/// an entire B in one shot or produce an error. It needs to be able to start the process for B,
/// then go somewhere, else, and then return and continue with B. This means that while initing B,
/// we need to have a state machine.
///
/// The state machine is advanced by passing it (&str, NextValueProducer) which it wants
/// It can be finalized when there are no more key-value pairs, producing Value or a set of errors.
#[doc(hidden)]
pub trait InitializationStateMachine<'de>: Sized {
    /// The target value being initialized by the machine
    type Value;

    /// True if the state machine would consume a key-value pair from serde map access if this key were available.
    fn wants_key(&self, key: &str) -> bool;

    /// Consume a key-value pair.
    fn next<NVP>(self, key: &str, next_value_producer: NVP) -> Self
    where
        NVP: NextValueProducer<'de>;

    /// Finalize self, producing either Self::Value or one or more errors.
    fn finalize(self) -> Result<Self::Value, Vec<InnerError>>;

    /// Check if this state machine needs to be finalized.
    ///
    /// In the majority of scenarios, this is ignored, and state machines are always finalized.
    ///
    /// For some features like flatten-optional, the field we need to initialize is `Option<Value>`,
    /// and we have a state machine for `Value` which we now need to decide whether to finalize.
    /// `needs_finalize` is queried to decide whether it should return `None` or attempt to finalize self
    /// and return `Some` (which may also fail and produce errors).
    ///
    /// In general, this should return true if this state machine consumed or would consume
    /// any values from non-serde sources, and false otherwise.
    fn needs_finalize(&self) -> bool;
}
```

This felt pretty clean and in testing, it works well in more exotic scenarios with multiple levels of flattening.

It also makes it easy to add support for "flatten with prefix" where all the fields can get prefixed with a string determined at the site of flattening.
